docs/src/pages/guides/typescript/typescript.md

```diff
@@ -247,3 +247,75 @@ import createMyTheme from './styles/createMyTheme';
 
 const theme = createMyTheme({ appDrawer: { breakpoint: 'md' }});
 ```
+
+## Usage of `component` property
+
+Material-UI allows you to replace a component's root node via a `component` prop.
+For example, a `Button`'s root node can be replaced with a React Router `Link`, and all extra props that are passed to `Button`, like `to`, will be spread down to the `Link` component, so you can do this: `<Button component={Link} to="/">Go Home</Button>`
+
+However, TypeScript will complain about this, because `to` is not part of the `ButtonProps` interface, and it has no way of inferring what props are what was passed in `component`.
+
+The current workaround is to spread these extra props down:
+
+```js
+interface LinkButtonProps extends ButtonProps {
+  to: LocationDescriptor,
+  replace?: boolean,
+}
+
+const renderLink = (props: any) => <Link {...props}/>
+
+const LinkButton = (props: LinkButtonProps) => (
+  <Button {...props} component={renderLink}/>
+)
+
+// usage:
+<LinkButton color="primary" to="/">Go Home</LinkButton>
+```
+
+Note that `renderLink`'s props param has a type of `any`.
+Material-UI components pass some basic event handler props (`onClick`, `onDoubleClick`, etc.) to their root nodes.
+These handlers have a signature of `(event: MouseEvent<HTMLElement, MouseEvent>) => void`, which is incompatible with the event handler signatures that `Link` expects, which are `(event: MouseEvent<AnchorElement>) => void`.
+
+Any element/component that you pass into `component` will have this problem if the signatures of their event handler props don't match.
+
+There is an ongoing effort to fix this by making component props generic.
+
+### Using React Router `Link` with `component` prop
+
+There are some other typing issues when using React Router's `Link` with the `component` property.
+
+TypeScript will throw an error if you try to pass `Link` in directly: `<ListItem component={Link}>`.
+Instead, `Link` should be wrapped in a functional component that is declared outside of the `render` before being passed in.
+
+> **NOTE:** There is currently an open issue that will require you to replace `ListItemLinkProps` with `any` in the signature for `createLink` below.
+
+To replace the root element of `ListItem`:
+
+```js
+import * as React from 'react'
+import { ListItem } from '@material-ui/core';
+import { Link } from 'react-router-dom';
+import { ListItemProps } from '@material-ui/core/ListItem';
+import { LocationDescriptor } from 'history';
+
+interface ListItemLinkProps extends ListItemProps {
+  to: LocationDescriptor
+  replace?: boolean
+}
+
+// Note: you must replace `ListItemLinkProps` with `any` here for the moment.
+const createLink = ({ innerRef, ...props }: ListItemLinkProps) => <Link {...props}>;
```

Comment:
CreateLink?

Reply:
I agree since it is used as a component: `<ListItem {...props} component={createLink}>`. Needs some further adjustment in the text we apply this change.